Psammead to Simorgh Component Migration [WSTEAMA-51]

pvaliani opened the pull request:
Resolves Jira Ticket WSTEAMA-51

**Overall change:**
- Migrates all Psammead components (unless deleted where stated) except for `@bbc/web-vitals` to a module called `legacy` in Simorgh
- Psammead components cannot be viewed in Simorgh storybook (at the moment) but Simorgh Storybook still works


**Code changes:**

- Removes `changelogs` and `package.json` files from each migrated component in `#legacy` module
- Removes `@bbc/psammead` `package.json` dependency entries from Simorgh's `package.json`
- Adds the following dependencies to Simorgh `package.json`:
  - `jalaali-js`
  - `xmldoc`
  - `octokit/rest`
  - `octokit/action`
  - `find-cache-dir`
  - `juggle/resize-observer`
  - `mkdirp`
  - `json5`
- Downgrades `moment` to  v2.24.0 and `moment-timezone` to v0.5.26 in Simorgh for compatibility with imported Psammead components
- Adds `babel/plugin-proposal-export-default-from` plugin and updates module exports to default export type to resolve component imports
- Removes irrelevent and unnecessary unit tests relating to the import/export of now unused `package.json` files within migrated psammead components
- Adds a `#legacy` alias to `dirAlias.js`, `jsconfig.json` and webpack config
- Adds a `watchOptions` entry with `ignored` using glob pattern to the `webpack config` which ignore the `tz` `moment-timezone-include` files. This caused webpack to continuously loop upon running the development server with `yarn dev`
- Replaces Psammead dependencies in `service configs` with absolute paths due to legacy folders due to lack of aliasing support 
- Removes unused `psammead-oembed` component and related `dompurify` dependency from Simorgh `package.json`
- Adds a glob pattern to `storybook` `main.js` to path to `legacy` so that Simorgh storybook still builds and functions correctly
- Updates snapshots to reflect new classnames generated for html elements due to migration of the components
- Updates `escapeRegex` within `psammead-media-player` to reflect Simorgh's `media-player`. Prevents regEx escapes on our domain and enables our GH actions CodeQL checks to pass

**Steps**:

 ✅ - **Completed**
❌ - **Not Completed**

Process *BEFORE* the start of the migration was set out and is described below:

- Copy all components from Psammead across to Simorgh into a 'legacy' folder in Simorgh to differentiate them from components we will move to in the future that meet our 'component standards' ✅
- Perform a mass find and replace to remove references to Psammead packages from Simorgh package.json - all "@bbc/psammead/ references, @bbc/gel-foundations replace with a #legacy alias ✅

- Then for each package we will bring it in line with how components are structured in Simorgh. We need to convert the packages from separate NPM packages to local components in Simorgh, this involved:
    - Deleting the package.json from each psammead component
    - hoisting any dependencies up to the root package.json in Simorgh ✅
    - Deleting changelog.md ✅
    - Deleting dist, esm, node_modules for individual package build artefacts (either do a fresh clone of Psammead without them or remove them from the legacy folder when copying over). ✅
Note: Any Psammead components that have NPM dependencies need to have their dependencies added to Simorgh package.json. For example:
https://github.com/bbc/psammead/blob/latest/packages/containers/psammead-timestamp-container/package.json which requires "moment-time". Hoist these dependencies to the top level of the Simorgh package.json ✅

- Consider moving content of src into root (move all source code up one level within the Psammead component folder name), and delete the now empty src directory - this may be essential, see if it breaks first ❌
- Check that the Psammead component tests run correctly - may need to update Simorgh test script ✅




**Additional work to be completed before merging this PR**:
- Investigate and reduce bundle sizes down to acceptable threshold. Largest bundle size is currently ~1300kB (https://github.com/bbc/simorgh/pull/10110) ✅ (see below)
- Perform change freeze (and probably merge freeze) and do a broad regression test across the site before putting live
Example (spike) PR: https://github.com/bbc/simorgh/pull/9966

**Additional work to be completed related to this PR at a future date**:
- Migrate `@bbc/web-vitals` from Psammead [WSTEAMA-79]
- Fix Psammead components to be viewable in Simorgh storybook [[NEWSWORLDSERVICE-1534]
- Decide the future of Psammead NPM packages and the Psammead repo [WSTEAMA-80]
- Upgrade Moment [NEWSWORLDSERVICE-1533] (https://github.com/bbc/simorgh/pull/10128) ✅

**BUNDLE SIZE FIX**

The Bundle Sizes have been addressed in this PR:

https://github.com/bbc/simorgh/pull/10110 

The largest bundle size is currently ~1090kB but further work should be completed to reduce this where possible

---

- [x] I have assigned myself to this PR and the corresponding issues
- [x] I have added the `cross-team` label to this PR if it requires visibility across World Service teams
- [ ] I have assigned this PR to the Simorgh project
- [x] (BBC contributors only) This PR follows the [repository use guidelines](https://github.com/bbc/simorgh-infrastructure/blob/latest/documentation/repository-guidelines.md)

**Testing:**

- [x] Automated (jest and/or cypress) tests added (for new features) or updated (for existing features)
- [ ] If necessary, I have run the local E2E non-smoke tests relevant to my changes (`CYPRESS_APP_ENV=local CYPRESS_SMOKE=false yarn test:e2e:interactive`)
- [x] This PR requires manual testing

pvaliani committed:
add legacy folder and copy over utilities and root readme

pvaliani committed:
add psammead-timestamp-container

pvaliani committed:
copies psammead components into legacy folder

pvaliani committed:
removes changelogs from legacy folder

pvaliani committed:
adds moment dependency from psammead timestamp container to simorgh package.json

pvaliani committed:
removes web-vitals from migration - to be done at a later date

pvaliani committed:
updates simorgh dependencies with psammead npm dependencies

pvaliani committed:
removes package.json from psammead components

pvaliani committed:
adds #legacy alias for imports

pvaliani committed:
replaces one path in gel foundations, adds babel export default plugin from psammead as dependency for Simorgh

pvaliani committed:
migrates gel-foundations - working in dev

pvaliani committed:
migrates psammead-styles component - working

pvaliani committed:
updates package.json - forgot to remove psammead-styles dependency

pvaliani committed:
migrates psammead-copyright - working in dev

pvaliani committed:
migrates psammead-useful-links - working in dev

pvaliani committed:
migrates psammead-script-link - working in dev

pvaliani committed:
migrates psammead-live-label - working in dev

pvaliani committed:
updates package.json removes live-label dependency

pvaliani committed:
migrates psammead-visually-hidden-text

pvaliani committed:
updates package.json -  removes psammead-visually-hidden-text

pvaliani committed:
adds #legacy alias to jsconfig to resolve module paths

pvaliani committed:
updates jsconfig incorrect legacy path - typo

pvaliani committed:
update alises in config

pvaliani committed:
migrates psammead-embed-error - working in dev

pvaliani committed:
migrates psammead-inline-link - working in dev

pvaliani committed:
migrates psammead-image-placeholder - working in dev

pvaliani committed:
migrates psammead-image - working in dev

pvaliani committed:
migrates psammead-paragraph - working in dev

pvaliani committed:
migrates psammead-figure - working in dev

pvaliani committed:
updates package.json to remove  psammead-figure

pvaliani committed:
migrates psammead-caption

pvaliani committed:
migrates psammead-section-label - working in dev

pvaliani committed:
migrates psammead-heading-index - working in dev

pvaliani committed:
migrates psammead-detokeniser - working in dev

pvaliani committed:
fixes psammead-heading-index migration error

pvaliani committed:
migrates psammead-byline - working in dev

pvaliani committed:
migrates psammead-play-button - working in dev

pvaliani committed:
migrates psammead-media-indicator - working in dev

pvaliani committed:
migrates psammead-bulletin - working in dev

pvaliani committed:
migrates psammead-grid - working in dev

pvaliani committed:
migrates psammead-topic-tags -working in dev

pvaliani committed:
migrates psammead-consent-banner - working in dev

pvaliani committed:
migrates psammead-calendars - working in dev

pvaliani committed:
migrates psammead-bulleted-list - working in dev

pvaliani committed:
migrates psammead-story-promo-list - working in dev

pvaliani committed:
migrates psammead-brand - re-routes skiplink and removes unused package.json

pvaliani committed:
migrates psammead-timestamp-container - working in dev

pvaliani committed:
migrates psammead-timestamp - working in dev

pvaliani committed:
migrates psammead-rich-text-transforms - working in dev

pvaliani committed:
migrates psammead-story-promo - re-routes IndexAlsos due to deleted package.json - working in dev

pvaliani committed:
migrates psammead-social-embed - working in dev

pvaliani committed:
migrates psammead-navigation - working in dev

pvaliani committed:
removes psammead-navigation package.json reference

pvaliani committed:
migrates psammead-asset + fixes issue with loadable causing psammead-rich-text-transforms to prevent MAP pages rendering - working in dev

DarioR01 committed:
increase bundle size (200+ increase)

DarioR01 committed:
install eslint-plugin-es5@latest in dev dependencies

DarioR01 committed:
fix all prettier errors

DarioR01 committed:
revert installation of eslint 5 plugin, and remove eslintrc using eslint 5 plugin

DarioR01 committed:
some fixes (need to fix psammead-locales/moment)

DarioR01 commented:
I might have accidentally changed the test logic of psammead-locales... All the files can be identified by running yarn test and looking for error 'i' is already declared in the upper scope on line 16 column 7

I approached this by removing the eslintrc file in psammead locales, but we could also install esling-plugin-es5 so that psammead-locales uses es5 instead.

pvaliani committed:
updates snapshots and psammead-rich-text-transforms to default export

pvaliani committed:
umerges dario updates with mine"

pvaliani committed:
removes unnecessary psammead-styles tests

pvaliani committed:
updates exports for failing candy-xml unit tests in psammead-rich-text-transforms

pvaliani committed:
removes unnecessary gel-foundations export tests on package.json"
"

pvaliani committed:
removes unnecessary psammead-assets export package.json unit tests

pvaliani committed:
migrates psammead-locales - working in dev

pvaliani committed:
downgrades moment and moment-timezone for psammead migration compatibility - tests were failing, now they're passing, woo hoo

pvaliani committed:
migrates moment-timezone-include - working in dev but still requires npm dependency - may need further review

pvaliani committed:
cleanup of empty folders and unused imports

pvaliani committed:
fixes lint errors, storybook preview imports and adds ignore pattern for legacy folder

RetiredBlob committed:
Merge branch 'latest' into WSTEAMA-51-psammead-component-migration

DarioR01 committed:
fix: webpack loop fix

DarioR01 committed:
fix: linting errors from yarn test

DarioR01 committed:
fix: remove @bbc/moment-timezone-include find-cache-dir object-hash unused packages

DarioR01 committed:
fix: dependencyCheck ignores stories.js files

ryanmccombe committed:
Merge branch 'latest' into WSTEAMA-51-psammead-component-migration

DarioR01 committed:
remove dependency test ignorePatterns and installs json5 dependency

DarioR01 committed:
fix: no-shadow linting, changed linting logic to ignore specific files/ folders

DarioR01 committed:
linting: massively reduction of linting warnings

DarioR01 committed:
remove all errors

DarioR01 committed:
Merge branch 'latest' into WSTEAMA-51-psammead-component-migration

DarioR01 committed:
fix: brandSVG imports

DarioR01 committed:
fix: script imports

DarioR01 committed:
fix: dario is a genius :D

pvaliani committed:
updates to service configs - removes aliases and replaces with absolute paths instead so cypress can run

pvaliani committed:
Merge branch 'latest' into WSTEAMA-51-psammead-component-migration

DarioR01 committed:
fix: some cypress tests

DarioR01 committed:
Merge branch 'latest' into WSTEAMA-51-psammead-component-migration

DarioR01 committed:
Revert "fix: some cypress tests"

This reverts commit 2ecda2c96b13a5e5f33124eee2ec9af0318ba7d2.

DarioR01 committed:
Merge branch 'latest' into WSTEAMA-51-psammead-component-migration

DarioR01 committed:
fix cypress tests

pvaliani committed:
update snapshots

pvaliani committed:
updates jalaali dependency in psammead-calendars

pvaliani committed:
storybook working again for simorgh - adds ignore glob pattern to legacy module

pvaliani committed:
update package json scripts to build simorgh and sync tz folder generation within moment-timezone include - prevents gh actions breaking

pvaliani committed:
removes unused component psammead-oembed and dompurify dependency associated with

pvaliani committed:
forgot 2 do a yarn

DarioR01 committed:
fix cypress again after accidentaly deleting our work :D

DarioR01 committed:
disable linting in cypress plugin file

DarioR01 committed:
Merge branch 'latest' into WSTEAMA-51-psammead-component-migration

DarioR01 committed:
fix regex escape

DarioR01 committed:
Merge branch 'WSTEAMA-51-psammead-component-migration' of github.com:bbc/simorgh into WSTEAMA-51-psammead-component-migration

DarioR01 committed:
regex fix secong trial

DarioR01 committed:
Third attempt at regex fix

DarioR01 committed:
remote branch is stuck, trying with another push

DarioR01 committed:
Merge branch 'latest' into WSTEAMA-51-psammead-component-migration

DarioR01 committed:
fourth attempt at fixing regex

DarioR01 committed:
Merge branch 'WSTEAMA-51-psammead-component-migration' of github.com:bbc/simorgh into WSTEAMA-51-psammead-component-migration

pvaliani committed:
matches regex escape handling in psammead to what is in current Simorgh for mediaplayer

pvaliani committed:
updates dynamic method call w codeql comment

DarioR01 committed:
bundles size greately reduced and calculated

pvaliani committed:
Merge branch 'latest' into WSTEAMA-51-psammead-component-migration

DarioR01 committed:
Revert "bundles size greately reduced and calculated"

This reverts commit a06ca9421c88dd62cdd29a195c0d31bfa9bb68ee.

DarioR01 committed:
fix: webpack bundle size

DarioR01 committed:
fix: revert imports in services

DarioR01 committed:
revert: webpack tests

pvaliani committed:
Merge branch 'latest' into WSTEAMA-51-psammead-component-migration

pvaliani committed:
Merge branch 'latest' into WSTEAMA-51-psammead-component-migration

amoore108 committed:
Changed imports of `psammead-assets/svgs`

- Changed import structure of SVGs, removing destructuring
- Renamed bbcBlocks and bbcBlocksDarkmode to `.jsx`
-

HarveyPeachey committed:
Revert "Changed imports of `psammead-assets/svgs`"

This reverts commit fa886fc0faf318f5cf59d444660b1cc6030ffa59.

HarveyPeachey committed:
adds side effect hinting, to reduce bundle size via file exports

HarveyPeachey committed:
Merge https://github.com/bbc/simorgh into WSTEAMA-51-psammead-migration-webpack

HarveyPeachey committed:
set bundle size limits to original values

pvaliani committed:
Merge branch 'latest' into WSTEAMA-51-psammead-component-migration

DarioR01 committed:
Merge branch 'latest' into WSTEAMA-51-psammead-component-migration

DarioR01 committed:
increase bundle size thresholds

DarioR01 committed:
Merge branch 'WSTEAMA-51-psammead-component-migration' of github.com:bbc/simorgh into WSTEAMA-51-psammead-migration-webpack

HarveyPeachey committed:
adds more specific side effect glob patterns

pvaliani committed:
Merge branch 'latest' into WSTEAMA-51-psammead-component-migration

HarveyPeachey committed:
changed allowlist

HarveyPeachey committed:
adds cypress to side effect list

DarioR01 committed:
Merge branch 'latest' into WSTEAMA-51-psammead-component-migration

pvaliani committed:
Merge pull request #10110 from bbc/WSTEAMA-51-psammead-migration-webpack

Fix Webpack Bundle Size for Psammead Migration [WSTEAMA-51]

pvaliani committed:
fix linting

pvaliani committed:
resolve conflicts

paruchurisilpa committed:
Merge branch 'latest' into WSTEAMA-51-psammead-component-migration

paruchurisilpa commented:
Testing:

Have regression tested links for few services on this document below locally -
https://github.com/bbc/simorgh/blob/latest/docs/Simorgh-Release-Info.md

Covered:
Chrome Latest
Safari on Last 2 IOS versions
Safari 13.x
Safari 14.x
Opera Mini (Android)
Including Extreme Mode
Latest Opera on Desktop
Latest Firefox on Desktop

paruchurisilpa commented:
Cypress tests seem to have passed too. Happy for this to be merged..

pvaliani committed:
Merge branch 'latest' into WSTEAMA-51-psammead-component-migration

pvaliani committed:
Merge branch 'latest' into WSTEAMA-51-psammead-component-migration

MuhammadAHussain commented on .eslintrc.js:
```diff
@@ -21,7 +21,8 @@ module.exports = {
     },
   },
-  plugins: ['prettier', 'json', 'jsx-a11y', 'react-hooks', 'cypress'],
+  ignorePatterns: ['**/tz/**', 'index.stories.jsx', 'index.amp.stories.jsx'],
```
Curious why we are ignoring story pages now?

pvaliani replied:
It's to do with the linting errors we are receiving at the moment due to Psammead components not being viewable in Simorgh. There are currently two separate `main.js` config files between the Simorgh and Psammead repo's which are used to configure their respective storybooks. A ticket has been setup and is detailed in the body of this PR description in order to address work required to get Psammead components viewable in Simorgh.

amoore108 commented on src/app/containers/ArticleTimestamp/timeFormatTests/expectedFormats.json:
```diff
@@ -101,10 +101,10 @@
   },
   "gujarati": {
     "default": {
-      "one minute ago": "એક મિનિટ પહેલા",
-      "five minutes ago": "5 મિનિટ પહેલા",
-      "one hour ago": "એક કલાક પહેલા",
-      "five hours ago": "5 કલાક પહેલા",
```
Looks like a slight difference in formatting here. May not be an issue as it seems to only affect this service.

pvaliani replied:
Can keep an eye on it! However, there's a PR and ticket in the body description to address the downgrading in this PR of Moment (which may be the culprit) and to re-upgrade it post-deployment which might solve any formatting discrepancies.

amoore108 reviewed (approved):
Looks good. Great work by everyone involved! 🚀

MuhammadAHussain reviewed (approved):
LGTM 👍

amoore108 committed:
Merge branch 'latest' into WSTEAMA-51-psammead-component-migration
